dlt/destinations/snowflake/sql_client.py

```diff
@@ -0,0 +1,164 @@
+from dlt.destinations.snowflake import capabilities
+
+
+class SnowflakeSqlClient(SqlClientBase[snowflake_lib.SnowflakeConnection], DBTransaction):
+
+    dbapi: ClassVar[DBApi] = snowflake_lib
+    capabilities: ClassVar[DestinationCapabilitiesContext] = capabilities()
+
+    def __init__(self, dataset_name: str, credentials: SnowflakeCredentials) -> None:
+        super().__init__(dataset_name)
+        self._conn: snowflake_lib.SnowflakeConnection = None
+        self.credentials = credentials
+
+    def open_connection(self) -> snowflake_lib.SnowflakeConnection:
+        self._conn = snowflake_lib.connect(
+            user=self.credentials.username,
+            password=self.credentials.password,
+            account=self.credentials.host,
+            warehouse=self.credentials.warehouse,
+            database=self.credentials.database,
+            schema=self.fully_qualified_dataset_name(),
+            role=self.credentials.role
+        )
+        return self._conn
+
+    @raise_open_connection_error
+    def close_connection(self) -> None:
+        if self._conn:
+            self._conn.close()
+            self._conn = None
+
+    @contextmanager
+    def begin_transaction(self) -> Iterator[DBTransaction]:
+        try:
+            self._conn.autocommit(False)
+            yield self
+            self.commit_transaction()
+        except Exception:
+            self.rollback_transaction()
+            raise
+
+    @raise_database_error
+    def commit_transaction(self) -> None:
+        self._conn.commit()
+        self._conn.autocommit(True)
+
+    @raise_database_error
+    def rollback_transaction(self) -> None:
+        self._conn.rollback()
+        self._conn.autocommit(True)
+
+    @property
+    def native_connection(self) -> "snowflake_lib.SnowflakeConnection":
+        return self._conn
+
+    def has_dataset(self) -> bool:
+        query = """
+                SELECT 1
+                    FROM INFORMATION_SCHEMA.SCHEMATA
+                    WHERE schema_name = %s;
+                """
+        rows = self.execute_sql(query, self.fully_qualified_dataset_name(escape=False))
+        return len(rows) > 0
+
+    def create_dataset(self) -> None:
+        self.execute_sql("CREATE SCHEMA %s" % self.fully_qualified_dataset_name())
+
+    def drop_dataset(self) -> None:
+        self.execute_sql("DROP SCHEMA %s CASCADE;" % self.fully_qualified_dataset_name())
+
+    def execute_sql(self, sql: AnyStr, *args: Any, **kwargs: Any) -> Optional[Sequence[Sequence[Any]]]:
+        with self.execute_query(sql, *args, **kwargs) as curr:
+            if curr.description is None:
+                return None
+            else:
+                f = curr.fetchall()
+                return f
+
+    @contextmanager
+    @raise_database_error
+    def execute_query(self, query: AnyStr, *args: Any, **kwargs: Any) -> Iterator[DBApiCursor]:
+        curr: DBApiCursor = None
```

Comment:
snowflake has a nice API to get dataframes and chunked dataframes from cursor. could you implement something like
```
class BigQueryDBApiCursorImpl(DBApiCursorImpl):
    """Use native BigQuery data frame support if available"""
    native_cursor: BQDbApiCursor  # type: ignore

    def df(self, chunk_size: int = None, **kwargs: Any) -> DataFrame:
        query_job: bigquery.QueryJob = self.native_cursor._query_job

        if chunk_size is None:
            try:
                return query_job.to_dataframe(**kwargs)
            except ValueError:
                # no pyarrow/db-types, fallback to our implementation
                return super().df()
        else:
            return super().df(chunk_size=chunk_size)
```

btw. this could also be used to wrap mutliple cursors returned from snowflake into something dbapi-like but I think it is a waste of time

Reply:
https://github.com/dlt-hub/dlt/pull/414/files#diff-36ef69a1f6bf9feca5e587f6594e04c9750f29d36f5a75d00fcefef7c7226ac0R13-R20  
Opted to keep the df columns in their original uppercase, not sure if better to lowercase them. This way it matches what's in the db

Reply:
we should mess with the resultsets that snowflake returns